quil-rs/src/program/mod.rs

```diff
@@ -104,6 +108,41 @@ impl Program {
             Instruction::WaveformDefinition(WaveformDefinition { name, definition }) => {
                 self.waveforms.insert(name, definition);
             }
+            Instruction::Gate(gate) => {
+                self.used_qubits.extend(gate.qubits.clone());
+                self.instructions.push(Instruction::Gate(gate));
+            }
+            Instruction::Measurement(measurement) => {
+                self.used_qubits.insert(measurement.qubit.clone());
+                self.instructions
+                    .push(Instruction::Measurement(measurement));
+            }
+            Instruction::Reset(reset) => {
+                if let Some(qubit) = &reset.qubit {
+                    self.used_qubits.insert(qubit.clone());
+                }
+                self.instructions.push(Instruction::Reset(reset));
+            }
+            Instruction::Delay(delay) => {
+                self.used_qubits.extend(delay.qubits.clone());
+                self.instructions.push(Instruction::Delay(delay));
+            }
+            Instruction::Fence(fence) => {
+                self.used_qubits.extend(fence.qubits.clone());
+                self.instructions.push(Instruction::Fence(fence));
+            }
+            Instruction::Capture(capture) => {
+                self.used_qubits.extend(capture.frame.qubits.clone());
+                self.instructions.push(Instruction::Capture(capture));
+            }
+            Instruction::Pulse(pulse) => {
+                self.used_qubits.extend(pulse.frame.qubits.clone());
+                self.instructions.push(Instruction::Pulse(pulse));
+            }
+            Instruction::RawCapture(raw_capture) => {
+                self.used_qubits.extend(raw_capture.frame.qubits.clone());
+                self.instructions.push(Instruction::RawCapture(raw_capture));
+            }
```

Comment:
Moved from `get_used_qubits` with the addition of adding each instruction to the instructions list as well. I decided this was better than

```rust
other => {
    match &other {
        // Add qubits to set
        _ => {
            // Has no qubits, ignore
        }
    }
    self.instructions.push(other);
}
```

because it makes it more difficult for a branch to slip through the cracks, e.g. a new instruction that should be handled specially but isn't caught by the compiler because it falls through both default cases.